### ISSUE TITLE
[HOLD for payment 2023-06-08] [$1000] Italic [Attachment] message is considered as attachment and app crashes on click of download for that message

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#17289]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app and login with user A
2. Open the app on another device and login with user B
3. From user A, send following message: `_[Attachment]_`
4. From user B, long press the message and observe that 'Download' is available in option in place of 'copy to clipboard'
5. Press on download

## Expected Result: 
App should not consider normal message as attachment
## Actual Result: 
App considers normal message as attachment
## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** v1.2.98-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/231275654-c349b140-9041-487a-abc0-e65b17aaa7de.mp4

https://user-images.githubusercontent.com/43996225/231275706-d2519f60-f513-413f-b0a0-a606ee1b2972.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681205583378899

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013ef972b4b00c3551</li>
        <li>Upwork Job ID: 1646266195415273472</li>
        <li>Last Price Increase: 2023-05-08</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @lschurr (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by lschurr** [comment]
I'm not able to reproduce this. I only have the "copy to clipboard" or "mark as unread" option when long holding on the message.

Will add Eng for more eyes.

**Comment by dangrous** [comment]
Oop yep @lschurr did you make the `[Attachment]` italic (i.e. `_[Attachment]_`)? That made it work for me. It's in the title but it's not in the reproduction instructions

**Comment by lschurr** [comment]
Ah, I'll change the repro steps. Do you think this should be external or internal @dangrous?

**Comment by dangrous** [comment]
I think External should be fine!

**Comment by MelvinBot** [comment]
Unable to auto-create job on Upwork. The BZ team member should create it manually for this issue. cc @thienlnam

**Comment by MelvinBot** [comment]
Current assignee @lschurr is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`External`)

**Comment by MelvinBot** [comment]
Current assignee @dangrous is eligible for the External assigner, not assigning anyone new.

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Send a italic [Attachment] message

```
_[Attachment]_
```

 It's considered as an attachment and can be downloaded, which is incorrect.

### What is the root cause of that problem?

The  italic [Attachment] message

```
_[Attachment]_
```

 is saved as 

```json
{
	"html": "<em>[Attachment]</em>",
  "text": "[Attachment]"
}
```

See below page inspection picture 

<img width="800" alt="image" src="https://user-images.githubusercontent.com/117511920/231368531-502195f2-aae7-4835-8380-e34711774dd1.png">

We check if a report action is an attachment here

https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/pages/home/report/ContextMenu/ContextMenuActions.js#L92-L94

If property `isAttachment` of  `reportAction` is set, then we use it to determine if this action is an attachment. If not, we check it through message using method [isReportMessageAttachment](https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/libs/isReportMessageAttachment.js#L10).

From the inspection picture above, the property `isAttachment` of action is set to `true` and `isAttachment` is correctly set from frontend method [buildOptimisticAddCommentReportAction](https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/libs/ReportUtils.js#L928). So the root cause is that backend sets the wrong value to property `isAttachment`.

### What changes do you think we should make in order to solve the problem?

To fix this issue, a backend fix will be needed to set right value to property `isAttachment`. Frontend fix is also necessary as there's a flaw in method `isReportMessageAttachment`

https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/libs/isReportMessageAttachment.js#L10-L12

The second condition `html !== CONST.ATTACHMENT_MESSAGE_TEXT` is not always correct for new comment like

```
_[Attachment]_
```

```
[[Attachment]](https://www.google.com)
```

Those comments' html are not equal to `CONST.ATTACHMENT_MESSAGE_TEXT` but they're considered as attachments though they're not attachments.

Searching usage of method `isReportMessageAttachment` in the codebase, we can find that, in some cases, we only use `message` to check if an action is an attachment. For example, the copy to clipboard icon won't be shown in the context menu after the backend fixing the invalid `isAttachment` issue, see

https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/pages/home/report/ContextMenu/ContextMenuActions.js#L141-L143

We can have a more robust `isReportMessageAttachment` method by reusing [this regex](https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/libs/fileDownload/getAttachmentDetails.js#L13) from `getAttachmentDetails.js` which contains a special attribute name of Expensify source. The new method will be like

```js
export default function isReportMessageAttachment({text, html}) {
    const regex = new RegExp(`${CONST.ATTACHMENT_SOURCE_ATTRIBUTE}*=*"(.+?)"`, 'i');
    return text === CONST.ATTACHMENT_MESSAGE_TEXT && html.match(regex);
}
```

We can also check `isAttachment` and `isReportMessageAttachment` together to determine if an action is an attachment. We can improve this kind of checking 

https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/pages/home/report/ContextMenu/ContextMenuActions.js#L92-L94

to 

```js
const isAttachment = lodashGet(reportAction, 'isAttachment', false) && ReportUtils.isReportMessageAttachment(message);
```

### What alternative solutions did you explore? (Optional)

N/A

**Comment by lschurr** [comment]
Job is here: https://www.upwork.com/jobs/~0156d17d888253de92

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~013ef972b4b00c3551

**Comment by MelvinBot** [comment]
Current assignee @lschurr is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @rushatgabhane is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @dangrous is eligible for the External assigner, not assigning anyone new.

**Comment by lschurr** [comment]
Looks like we have a few proposals already. @rushatgabhane could you review?

**Comment by eh2077** [comment]
### Proposal
[Updated](https://github.com/Expensify/App/issues/17289#issuecomment-1505972593)

**Comment by lschurr** [comment]
Any update on this one @rushatgabhane?

**Comment by dangrous** [comment]
Just hopping in here - I agree this will need to be a front and backend fix together. Good news is I think we should be able to use pretty much the same logic so I can steal whatever we decide on for the front end and apply to the back. Using the text by itself to figure out whether or not there's an attachment seems... uh... bad? But using something like `CONST.ATTACHMENT_SOURCE_ATTRIBUTE` as proposed is definitely better, if there's not some other way to guarantee an attachment is an attachment.

Let me know what you think of the proposals, @rushatgabhane!

**Comment by eh2077** [comment]
Hi @rushatgabhane , thanks for checking this issue! 

I think frontend fix is necessary because, in the App codebase, we're using method isReportMessageAttachment only some places to identify the attachment, like
https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/pages/home/report/ContextMenu/ContextMenuActions.js#L141-L143
As there's a flaw in method isReportMessageAttachment, so the backend fix is not enough to fix all relevant issues. My original proposal https://github.com/Expensify/App/issues/17289#issuecomment-1505972593 covered more details about this.

**Comment by eh2077** [comment]
@rushatgabhane 

Here's the method isReportMessageAttachment
https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/libs/isReportMessageAttachment.js#L10-L12


The second condition html !== CONST.ATTACHMENT_MESSAGE_TEXT is not always correct for new comment _[Attachment]_
```
{
	"html": "<em>[Attachment]</em>",
        "text": "[Attachment]"
}
```

**Comment by eh2077** [comment]
> yep that's true but we'll never get to it because we're checking the backend response first. report.isAttachment
> 
> Please correct me if im wrong

Yes, we're supposed to check the backend response report.isAttachment first and then check isReportMessageAttachment everywhere consistently in the codebase. But, actually, we're not and, in many places, we only check it through method isReportMessageAttachment, like here
https://github.com/Expensify/App/blob/8adee6bdf733058985e55ba3fc0d39f72cddc5cf/src/pages/home/report/ContextMenu/ContextMenuActions.js#L141-L143

You can search usages of method isReportMessageAttachment in the codebase to find other cases

**Comment by dangrous** [comment]
Hm @Prince-Mendiratta if you can figure that out, let us know!

Assuming that `isAttachment` always exists (which we think it doesn't), I think the ideal solution would be to:
* Update the backend to more appropriately check for attachments - probably using `CONST.ATTACHMENT_SOURCE_ATTRIBUTE` until we have real placeholders
* Update the frontend to check for isAttachment in all places (such as the ones @eh2077 called out) instead of its own logic.

However, knowing that isAttachment isn't quite consistent, I'm not entirely sure what the best way forward is. If we can't identify (and fix) where that's happening, I think looking for `CONST.ATTACHMENT_SOURCE_ATTRIBUTE` both front and back end is probably our best bet for now.

**Comment by eh2077** [comment]
@rushatgabhane I sorted out many places in the App codebase that use method isReportMessageAttachment only below FYI

https://github.com/Expensify/App/blob/f54a8bd23018644d33114d1820f7e428f18c74f9/src/libs/OptionsListUtils.js#L408

https://github.com/Expensify/App/blob/f54a8bd23018644d33114d1820f7e428f18c74f9/src/libs/ReportActionsUtils.js#L148

https://github.com/Expensify/App/blob/f54a8bd23018644d33114d1820f7e428f18c74f9/src/libs/ReportUtils.js#L148

https://github.com/Expensify/App/blob/f54a8bd23018644d33114d1820f7e428f18c74f9/src/libs/SidebarUtils.js#L262

**Comment by eh2077** [comment]
@rushatgabhane Thanks for your patience! I just updated comment https://github.com/Expensify/App/issues/17289#issuecomment-1513952262 for your information

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by dangrous** [comment]
Okay backend fix is in process! We'll definitely need some front end fixes still for at least that ContextMenuActions stuff (when I was testing, I no longer saw the download icon for `_[Attachment]_` but I ALSO didn't see the edit message.

Also, I realized that using the string in `CONST.ATTACHMENT_SOURCE_ATTRIBUTE` isn't *perfect*, as you can still fool it by writing `[[Attachment]](https://www.google.com/data-expensify-source)` or something else that manually includes that phrase in the html, but I think we're much closer and you're getting into really weird edge cases at that point.

**Comment by eh2077** [comment]
> Also, I realized that using the string in `CONST.ATTACHMENT_SOURCE_ATTRIBUTE` isn't _perfect_, as you can still fool it by writing `[[Attachment]](https://www.google.com/data-expensify-source)` or something else that manually includes that phrase in the html, but I think we're much closer and you're getting into really weird edge cases at that point.

That's true and I also pointed out this in my proposal https://github.com/Expensify/App/issues/17289#issuecomment-1505972593 and suggested to use a regex to check if there's the special property inside an anchor tag.

**Comment by dangrous** [comment]
Cool! That's in review, with the update to the regex since that does seem to catch more instances. Not sure it's foolproof but we'll take what we can get. At this point someone would really have to be *trying* to get their app to break.

**Comment by MelvinBot** [comment]
@dangrous, @rushatgabhane, @lschurr Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by lschurr** [comment]
What's the latest on this one @dangrous @rushatgabhane?

**Comment by dangrous** [comment]
Backend fix is still in review (I'll ping again); we will still need some frontend updates, I wonder if we can sort of gauge the proposals based on the fact that we will "know" that `isAttachment` will be correct, and then see what other changes are needed?

**Comment by MelvinBot** [comment]
@dangrous @rushatgabhane @lschurr this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by dangrous** [comment]
Backend fix still being reviewed (we had to do a reviewer swap coz of some dev environment issues). Should be soon! I'll actually put this on HOLD until that's up, just to make things clear

**Comment by dangrous** [comment]
Yep - the backend fix is merged but not yet deployed. This should be pretty quick to accept a proposal and close out once that's ready to go.

**Comment by dangrous** [comment]
Almost there! We're on staging

**Comment by dangrous** [comment]
Uh Melvin chill please.

**Comment by dangrous** [comment]
That backend fix is now on prod so we're good to move forward! Do we need to update any of the proposals based on that face (and/or @rushatgabhane was [this](https://github.com/Expensify/App/issues/17289#issuecomment-1536191010) you picking/reviewing @eh2077's proposal?)

Thanks!

**Comment by lschurr** [comment]
@rushatgabhane could you look this one over and let me know what next steps we need to take?

**Comment by eh2077** [comment]
@Prince-Mendiratta @rushatgabhane 

One fact is that both of us submitted the original proposals by using a bot. So there's evidence that we both _**independently**_ wrote the RCA of the issue before the `Help Wanted` tag was added. 

While I think I dug deeper about RCA and came up a more solid solution by using regex which is adopted by @dangrous for the backend fix. So I think my proposal should be compensated with the full bounty of this issue.

But I'd also respect the decision made by C+ team.

cc @dangrous @lschurr

**Comment by esh-g** [comment]
Wow it happened again 😅 @eh2077

**Comment by dangrous** [comment]
Hey! So I think since both of you were active in the discussion to get this sorted and identified, it might make the most sense to split. BUT, @eh2077 is right that I stole their regex (😄 ) so is it weird to hire both and do like a $750/$250 split?

@lschurr lemme know how that sounds to you.

In terms of raising an actual PR, @eh2077 would you be willing to do that side of it?

**Comment by eh2077** [comment]
> Hey! So I think since both of you were active in the discussion to get this sorted and identified, it might make the most sense to split. BUT, @eh2077 is right that I stole their regex (😄 ) so is it weird to hire both and do like a $750 eh/$250 prince split?

@dangrous Thanks for your input! I also agree it's weird to share the bounty because the fact is that I **_independently_** wrote the RCA and proposed the better solution. Note as I mentioned before both of our original proposals are submitted by using a bot/script, which means we worked in advance on the issue before it was tagged with `Help Wanted` label.

Also based on the [proposal selection rule](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#check-github-for-existing-proposals-from-other-users)

> Expensify reviews all solution proposals on a first come first serve basis. If you see other contributors have already proposed a solution, you can still provide a solution proposal and we will review it. We look for the earliest provided, best proposed solution that addresses the job.

I think, in this case, **_the best proposal should be compensated with the whole bounty of this issue._**. 

If the team also decides to compensate @Prince-Mendiratta a special bonus(not deducted from the $1000 bounty of this issue), like $250, for his work on investigating this issue, I'd love to see that👍


> In terms of raising an actual PR, @eh2077 would you be willing to do that side of it?

Yes, sure I'm willing to raising a PR for this issue.


@dangrous @lschurr

**Comment by dangrous** [comment]
Okay yeah I think that's fair, let's proceed for now with @eh2077's PR and full payment, and we can figure out the rest later. Will assign you now!

**Comment by eh2077** [comment]
@dangrous @rushatgabhane The PR https://github.com/Expensify/App/pull/18766 is ready. Please help to review it at your convenience, thanks!

**Comment by lschurr** [comment]
@dangrous @rushatgabhane looks like the PR is still waiting for review.

**Comment by lschurr** [comment]
How are we doing on this one @dangrous @rushatgabhane?

**Comment by dangrous** [comment]
Just merged the PR! Should get the automessages here soon

**Comment by lschurr** [comment]
I'm heading OOO from May 17-23rd. I've re-added the Bug label to re-assign the issue and kept myself as assigned as well. If this is still open on May 24th, I'll take it back from you @joekaufmanexpensify 

Latest update: PR just merged, waiting for auto-messages for BZ checklist/payment.

**Comment by lschurr** [comment]
Removing you @joekaufmanexpensify since I'm back from OOO.

**Comment by lschurr** [comment]
The job expired for this one, so I created a new one.

@eh2077 @rushatgabhane @dhanashree-sawant could you apply here? https://www.upwork.com/jobs/~018d53d5272968557e

**Comment by dhanashree-sawant** [comment]
Hi @lschurr, thanks I have just applied.

**Comment by eh2077** [comment]
@lschurr Applied. Thanks!

**Comment by s77rt** [comment]
I think the linked PR here caused a regression https://github.com/Expensify/App/issues/19379. Do you want to fix this as a follow up?

**Comment by eh2077** [comment]
I commented in the linked issue https://github.com/Expensify/App/issues/19379#issuecomment-1562067941

**Comment by dangrous** [comment]
Oh that's a weird regression. I mean weird that we coded it that way in the first place. Tricksy.

Given that we're within the regression period we should figure out that first and then we can get you sorted, @Prince-Mendiratta

**Comment by lschurr** [comment]
@dangrous - since there was a regression, are we holding on paying this one?

**Comment by s77rt** [comment]
@rushatgabhane I think we can fix that. And I think the "Uploading attachment..." for download buttons is something that is happening on main already. We can fix that as well. I mentioned both concerns (edit & download) [here](https://github.com/Expensify/App/issues/19379#issuecomment-1562017902).

I was thinking something like:

**isReportMessageAttachment**
```js
    return text === CONST.ATTACHMENT_MESSAGE_TEXT && (!!html.match(regex) || html === 'Uploading attachment...');
```

**ContextMenuActions shouldShow download**
```js
    return isAttachment && message.html !== 'Uploading attachment...' && reportAction.reportActionID;
```

**Comment by eh2077** [comment]
@s77rt Thanks for the suggestion to hide download button!

@dangrous @rushatgabhane I just submitted a PR https://github.com/Expensify/App/pull/19618  to fix the special regression. I update the test case and the attached demo video about the fix. Please let me know if you're both happy to have the fix! If you are, then please help to review the PR. Thank you!

**Comment by lschurr** [comment]
@dangrous @rushatgabhane are you able to review the PR that @eh2077 submitted?

**Comment by dangrous** [comment]
we're waiting on one update but then should be good to merge!

**Comment by eh2077** [comment]
@dangrous I guessed if we decided to define that string as a constant, then I think we should also support localization for it, like translating it into Spanish as well. Do you agree with this?

**Comment by eh2077** [comment]
> Umm no Because that text is added by backend as a placeholder for image

@rushatgabhane Thanks for the information! I'll save this context in the comment.

**Comment by eh2077** [comment]
@dangrous @rushatgabhane Please help to review PR https://github.com/Expensify/App/pull/19618, thank you!

**Comment by dangrous** [comment]
merged!

**Comment by lschurr** [comment]
@dhanashree-sawant @rushatgabhane @eh2077 - Could you apply for the job in Upwork?
https://www.upwork.com/jobs/~01d01935e7c40d4d46

**Comment by lschurr** [comment]
Oh shoot, you're right @eh2077 - got lost on the comments on this one. I will close https://www.upwork.com/jobs/~01d01935e7c40d4d46

**Comment by lschurr** [comment]
Looks like we just need @rushatgabhane to apply for this one - https://www.upwork.com/jobs/~018d53d5272968557e

**Comment by dhanashree-sawant** [comment]
Hi @lschurr, I already have a offer for the job on upwork.

**Comment by lschurr** [comment]
@rushatgabhane - please apply here: https://www.upwork.com/jobs/~018d53d5272968557e

**Comment by lschurr** [comment]
Also @rushatgabhane - do we need a regression test?

**Comment by dangrous** [comment]
Oh right - @lschurr we were going to pay @Prince-Mendiratta a bit of a bonus for helping out, I think $250 should be okay if that's alright with you?

**Comment by lschurr** [comment]
Cool! @Prince-Mendiratta can you apply here? https://www.upwork.com/jobs/~018d53d5272968557e

**Comment by lschurr** [comment]
Cool, all paid except for @rushatgabhane - once you accept the offer, we can pay and close this one.

**Comment by lschurr** [comment]
These are all paid. Closing.

**Comment by eh2077** [comment]
Hi @lschurr , can you help to close/end the UpWork contract of this job https://www.upwork.com/jobs/~018d53d5272968557e?
